### PR TITLE
Acutally use the permission potency and handle permissions correctly

### LINKED ISSUE #356: Permissions ending with ".*" seem to be not working with a negative potency

### PR DESCRIPTION
**Pull request by derklaro**
This pull request includes:

- [ ] breaking changes
- [X] no breaking changes

### Changes made to the repository:

The permission system now actually handles the potency of a permission and competes which permission is the correct one (with the highest priority) to use.

### Documentation of test results:

The already existing compile-time test still passes as well as the ingame test.

### Related issues/discussions:

Fixes #356

### CODE REVIEW

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/DefaultPermissionManagement.java [hunk 1]
```diff
@@ -85,7 +86,7 @@ public boolean testPermissible(@Nullable IPermissible permissible) {
     @Override
     @NotNull
     public Collection<IPermissionGroup> getGroups(@Nullable IPermissible permissible) {
-        Collection<IPermissionGroup> permissionGroups = new ArrayList<>();
+        List<IPermissionGroup> permissionGroups = new ArrayList<>();
```
Comment (TheMeinerLP) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/DefaultPermissionManagement.java:95]: Reduce of creation of new arrays in the memory. 
Try with clear and a class based array. 
Java saves for you the reference.

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/DefaultPermissionManagement.java [hunk 2]
```diff
@@ -103,6 +104,7 @@ public boolean testPermissible(@Nullable IPermissible permissible) {
             permissionGroups.add(this.getDefaultPermissionGroup());
         }
 
+        permissionGroups.sort(Comparator.comparingInt(IPermissionGroup::getPotency).reversed());
```
Comment (TheMeinerLP) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/DefaultPermissionManagement.java]: Make for this a class variable/constant to reuse comparator.

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/DefaultPermissionManagement.java [hunk 3]
```diff
@@ -120,15 +122,72 @@ public PermissionCheckResult getPermissionResult(@NotNull IPermissible permissib
...
+            } else {
+                return PermissionCheckResult.fromPermission(bestGroupMatch.compareTo(bestMatch) > 0 ? bestGroupMatch : bestMatch);
+            }
+        }
+        return PermissionCheckResult.DENIED;
     }
 
+    private Permission tryExtendedGroups(Permission permission, IPermissible permissible, String currentGroup, Set<String> testedGroups) {
+        Permission bestResult = null;
+        for (IPermissionGroup group : this.getGroups(permissible)) {
```
Comment (TheMeinerLP) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/DefaultPermissionManagement.java:159]: Replace wirh a while loop and iterator based on the best result boolean. To reduce for/while-loop operations

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/IPermissible.java [hunk 4]
```diff
@@ -247,30 +247,57 @@ default boolean addPermission(@NotNull String group, @NotNull String permission,
...
+     * against each other to find the permission with the highest positive or negative potency.
+     *
+     * @param permissions the permission to check for.
+     * @param permission  the initial permission to check against.
+     * @return the logic permission which has the highest potency.
+     */
+    default @Nullable Permission findMatchingPermission(@NotNull Collection<Permission> permissions, @NotNull Permission permission) {
+        Permission lastMatch = null;
+        // search for a better match
         for (Permission permissionEntry : permissions) {
```
Comment (TheMeinerLP) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/IPermissible.java:264]: Also here, replace with a iterator and a right conndition

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/IPermissible.java [hunk 5]
```diff
@@ -247,30 +247,57 @@ default boolean addPermission(@NotNull String group, @NotNull String permission,
...
+     * @param input the permission to correct.
+     * @return a new corrected instance if this permissible has a higher absolute potency, else {@code input}.
+     */
+    default @NotNull Permission correctPermission(@NotNull Permission input) {
+        if (Math.abs(input.getPotency()) >= Math.abs(this.getPotency())) {
+            // if the permission potency is higher as the permissible one it will override it
+            return input;
+        } else {
+            // the user potency is higher than the permission one (absolute) so the permissible potency should be used
+            return new Permission(input.getName(), this.getPotency(), input.getTimeOutMillis());
```
Comment (TheMeinerLP) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/permission/IPermissible.java]: Why we need here a new object?

File: cloudnet-modules/cloudnet-cloudperms/src/main/java/de/dytanic/cloudnet/ext/cloudperms/bukkit/BukkitCloudNetCloudPermissionsPermissible.java [hunk 6]
```diff
@@ -37,23 +36,26 @@ public BukkitCloudNetCloudPermissionsPermissible(Player player, IPermissionManag
...
     public Set<PermissionAttachmentInfo> getEffectivePermissions() {
         Set<PermissionAttachmentInfo> infos = new HashSet<>();
+
         IPermissionUser permissionUser = CloudNetDriver.getInstance().getPermissionManagement().getUser(this.player.getUniqueId());
-        if (permissionUser == null) {
-            return infos;
-        }
+        if (permissionUser != null) {
+            for (String group : Wrapper.getInstance().getServiceConfiguration().getGroups()) {
+                CloudNetDriver.getInstance()
```
Comment (TheMeinerLP) [inline cloudnet-modules/cloudnet-cloudperms/src/main/java/de/dytanic/cloudnet/ext/cloudperms/bukkit/BukkitCloudNetCloudPermissionsPermissible.java:43]: Is necessary to loop over all entries ?

File: cloudnet-modules/cloudnet-cloudperms/src/main/java/de/dytanic/cloudnet/ext/cloudperms/bukkit/BukkitCloudNetCloudPermissionsPermissible.java [hunk 7]
```diff
@@ -81,44 +83,55 @@ public boolean hasPermission(@NotNull String inName) {
...
+            if (result != PermissionCheckResult.DENIED) {
+                return result.asBoolean();
             }
         }
         return false;
     }
 
-    private boolean testParents(String inName, Permission parent, Permission lastParent, Predicate<Permission> parentAcceptor) {
+    private PermissionCheckResult testParents(String inName, Permission parent, Permission lastParent, Function<Permission, PermissionCheckResult> parentAcceptor) {
         for (Map.Entry<String, Boolean> entry : parent.getChildren().entrySet()) {
```
Comment (TheMeinerLP) [inline cloudnet-modules/cloudnet-cloudperms/src/main/java/de/dytanic/cloudnet/ext/cloudperms/bukkit/BukkitCloudNetCloudPermissionsPermissible.java:118]: make a better conndition for the for loop, for ex. use a iterator. Prevent a hard break with return

**Review by TheMeinerLP — COMMENTED** [review]
All in all, looks fine and clean

**Review by juliarn — APPROVED** [review]

**Pull request merged**